expression/collation.go

```diff
@@ -189,12 +168,103 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
 	return CoercibilityImplicit
 }
 
+func deriveCollation(ctx sessionctx.Context, funcName string, args []Expression, retType types.EvalType, argTps ...types.EvalType) (dstCharset, dstCollation string, coercibility Coercibility, err error) {
+	switch funcName {
+	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Reverse, ast.Upper, ast.Quote, ast.Coalesce:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+	case ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0])
+	case ast.InsertFunc:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0], args[3])
+	case ast.Lpad, ast.Rpad:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[0], args[2])
+	case ast.Elt, ast.ExportSet, ast.MakeSet:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args[1:]...)
+	case ast.FindInSet:
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+	case ast.Field:
+		if argTps[0] == types.ETString {
+			return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+		}
+	case ast.Locate, ast.Instr:
```

Comment:
How about `LOCATE(substr,str,pos)`?

Reply:
update, `locate` and `instr` should only use the first 2 arguments